File: easy/83_Remove_Duplicates_from_Sorted_List/main.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        root = head
        prev = head
        curr = head.next
        
        while curr:
            if prev.val == curr.val:
                curr = curr.next
                if not curr:
                    prev.next = None
            else:
                prev.next = curr
                prev = curr
                curr = curr.next

        return head

File: easy/83_Remove_Duplicates_from_Sorted_List/main.cpp

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    // // Solution with dealocatation
    // ListNode* deleteDuplicates(ListNode* head) {
    //     if (head == nullptr || head->next == nullptr) {
    //         return head;
    //     }

    //     ListNode* root = head;  
    //     ListNode* prev = head;
    //     ListNode* curr = head->next;

    //     while (curr != nullptr) {
    //         if (curr->val == prev->val) {
    //             ListNode* toDelete = curr;
    //             curr = curr->next;
    //             delete toDelete;  
    //             prev->next = curr; 

    //             if (curr == nullptr) {
    //                 prev->next = nullptr; 
    //             }
    //         } else {
    //             prev = curr;
    //             curr = curr->next;
    //         }
    //     }

    //     return root; 
    // }

    // Solution Without dealocatation
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == nullptr  ||  head->next == nullptr){
            return head;
        }

        ListNode* root = head;
        ListNode* prev = head;
        ListNode* curr = head->next;

        while(curr != nullptr){
            if(curr->val == prev->val){
                curr = curr->next;
                if(curr == nullptr){
                    prev->next = nullptr;
                }
            }
            else{
                prev->next = curr;
                prev = curr;
                curr = curr->next;
            }
        }
        return root;
    }
};